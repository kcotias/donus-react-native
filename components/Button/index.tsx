import * as React from 'react';
import { Pressable, Text, View, ActivityIndicator, TextStyle } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import styles from './styles';

interface ButtonProps {
  iconName?: string;
  title?: string;
  titleColor?: string;
  iconColor?: string;
  onPress: () => void;
  style?: any;
  textStyle?: TextStyle;
  isLoading?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  iconName,
  iconColor,
  title,
  titleColor,
  onPress,
  style,
  textStyle,
  isLoading,
  testID,
}: ButtonProps) => (
  <Pressable
    testID={testID}
    onPress={onPress}
    style={({ pressed }) => [{ opacity: pressed ? 0.5 : 1 }]}>
    <View style={style || { ...styles.container, paddingHorizontal: !iconName ? 20 : 0 }}>
      {iconName && <Ionicons name={iconName} size={28} color={iconColor || 'white'} />}
      {title && !isLoading && (
        <Text style={textStyle || { ...styles.title, color: titleColor || 'white' }}>{title}</Text>
      )}
      {isLoading && <ActivityIndicator color="#fff" />}
    </View>
  </Pressable>
);

export default React.memo(Button);
