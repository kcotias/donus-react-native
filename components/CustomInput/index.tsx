import React from 'react';
import { TextInput } from 'react-native';
import styles from './styles';

interface CustomInputProps {
  onChangeText: (text: string) => void;
  placeholder: string;
}

const CustomInput: React.FC<CustomInputProps> = (props: CustomInputProps) => (
  <TextInput
    {...props}
    onChangeText={props.onChangeText}
    placeholder={props.placeholder}
    style={styles.textInput}
    placeholderTextColor="gray"
  />
);

export default CustomInput;
