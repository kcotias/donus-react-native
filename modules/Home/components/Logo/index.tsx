import React from 'react';
import { View } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import styles from './styles';

const Logo: React.FC = () => (
  <>
    <View style={styles.smallOpaqueCircle} />
    <View style={styles.largeOpaqueCircle} />
    <View style={styles.logoOuterCircle}>
      <View style={styles.logoInnerCircle}>
        <Ionicons name="beer-outline" color="#ffe135" size={50} />
      </View>
    </View>
  </>
);

export default React.memo(Logo);
