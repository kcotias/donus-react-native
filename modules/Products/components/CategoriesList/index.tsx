import React from 'react';
import { ScrollView, Text, Pressable } from 'react-native';
import SkeletonContent from 'react-native-skeleton-content';
import { Colors } from '../../../../constants';
import styles from './styles';

type Category = {
  id: number;
  title: string;
};
interface CategoriesListProps {
  activeButton: number;
  onButtonPress: (id: number) => void;
  categories: Category[];
  isLoading: boolean;
}

const CategoriesList: React.FC<CategoriesListProps> = ({
  activeButton,
  onButtonPress,
  categories,
  isLoading,
}: CategoriesListProps) => {
  return (
    <ScrollView style={styles.container} horizontal showsHorizontalScrollIndicator={false}>
      {!isLoading &&
        categories?.map((item: Category, index: number) => {
          const isActive = item.id === activeButton;

          return (
            <Pressable
              key={item.id}
              style={[
                styles.buttonContainer,
                {
                  backgroundColor: isActive ? Colors.pallete.secondary : '#fff',
                  marginLeft: index === 0 ? 16 : 0,
                },
              ]}
              onPress={() => onButtonPress(item.id)}>
              <Text style={[styles.title, { color: isActive ? '#fff' : Colors.pallete.secondary }]}>
                {item.title}
              </Text>
            </Pressable>
          );
        })}
      {isLoading && (
        <SkeletonContent
          containerStyle={{ flexDirection: 'row' }}
          layout={[{}, {}, {}, {}].map((item, index) => {
            return {
              key: index,
              ...styles.buttonContainer,
              marginLeft: index === 0 ? 16 : 0,
            };
          })}
          isLoading
        />
      )}
    </ScrollView>
  );
};

export default React.memo(CategoriesList);
