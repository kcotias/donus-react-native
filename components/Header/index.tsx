import * as React from 'react';
import { View, Text } from 'react-native';
import Button from '../Button';
import styles from './styles';

interface HeaderProps {
  hasBackButton?: boolean;
  onBackPress?: () => void;
  hasTitle?: boolean;
}

const Header: React.FC<HeaderProps> = ({ hasBackButton, hasTitle, onBackPress }: HeaderProps) => (
  <View style={styles.container}>
    <View>{hasBackButton && <Button onPress={onBackPress} iconName="chevron-back-outline" />}</View>
    {hasTitle && <Text style={styles.title}>Tasty Drinks</Text>}
  </View>
);

export default React.memo(Header);
