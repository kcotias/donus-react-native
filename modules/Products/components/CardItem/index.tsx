import React, { useState } from 'react';
import { View, Text, Image } from 'react-native';
import { Button } from '../../../../components';
import { Colors } from '../../../../constants';
import { Product } from '../../../../types';
import styles from './styles';

interface CardItemProps {
  item: Product;
}

const CardItem: React.FC<CardItemProps> = ({ item }: CardItemProps) => {
  const [counter, setCounter] = useState(0);

  function onCounterPress(operation: string) {
    if (operation === 'increase') {
      setCounter(counter + 1);
    } else {
      setCounter(counter - 1);
    }
  }

  return (
    <View style={styles.cardContainer}>
      <View style={styles.textArea}>
        <Text ellipsizeMode="tail" numberOfLines={2} style={styles.itemTitle}>
          {item.title}
        </Text>
        <Text style={styles.itemPrice}>R$ {item.productVariants[0].price}</Text>
        <View style={styles.counterArea}>
          {counter > 0 && (
            <>
              <Button
                testID="decrease"
                onPress={() => onCounterPress('decrease')}
                iconName="remove-outline"
                iconColor={Colors.pallete.primaryDark}
              />
              <Text testID="counter" style={{ marginHorizontal: 3 }}>
                {counter}
              </Text>
            </>
          )}
          <Button
            testID="increase"
            onPress={() => onCounterPress('increase')}
            iconName="add-outline"
            iconColor={Colors.pallete.primaryDark}
          />
        </View>
      </View>
      <View style={styles.imgArea}>
        <Image resizeMode="contain" source={{ uri: item.images[0].url }} style={styles.img} />
      </View>
    </View>
  );
};

export default CardItem;
