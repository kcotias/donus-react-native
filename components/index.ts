import Header from './Header';
import Button from './Button';
import CustomInput from './CustomInput';

export { Header, Button, CustomInput };
