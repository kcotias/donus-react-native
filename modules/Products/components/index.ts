import CategoriesList from './CategoriesList';
import CardItem from './CardItem';

export { CategoriesList, CardItem };
