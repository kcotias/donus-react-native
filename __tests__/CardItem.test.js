import 'react-native';
import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { expect, it } from '@jest/globals';
import { CardItem } from '../modules/Products/components';

const productMock = {
  title: 'titleTest',
  productVariants: [{ price: 1 }],
  images: [{ url: 'imageurl' }],
};

describe('<CardItem/> Component', () => {
  it('matches snapshot', async () => {
    const { unmount, toJSON } = render(<CardItem item={productMock} />);

    expect(toJSON()).toMatchSnapshot();
    await unmount();
  });

  it('renders correctly and buttons work', async () => {
    const { getByTestId, getByText, unmount } = render(<CardItem item={productMock} />);

    expect(getByText(productMock.title)).toBeDefined();
    expect(getByText(`R$ ${productMock.productVariants[0].price}`)).toBeDefined();

    const increase = getByTestId('increase');
    expect(increase).toBeDefined();
    fireEvent.press(increase);

    const decrease = getByTestId('decrease');
    expect(decrease).toBeDefined();

    expect(getByText('1')).toBeDefined();
    fireEvent.press(increase);
    expect(getByText('2')).toBeDefined();
    fireEvent.press(decrease);
    expect(getByText('1')).toBeDefined();
    await unmount();
  });
});
