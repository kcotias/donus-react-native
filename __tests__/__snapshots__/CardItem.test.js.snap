// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardItem/> Component matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "borderRadius": 20,
      "elevation": 5,
      "flexDirection": "row",
      "height": 120,
      "marginBottom": 20,
      "marginHorizontal": 16,
      "padding": 15,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.25,
      "shadowRadius": 3.84,
    }
  }
>
  <View
    style={
      Object {
        "width": "65%",
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
      style={
        Object {
          "color": "#152130",
          "fontSize": 16,
          "fontWeight": "600",
          "marginBottom": 4,
        }
      }
    >
      titleTest
    </Text>
    <Text
      style={
        Object {
          "color": "green",
          "fontSize": 18,
          "fontWeight": "600",
          "marginBottom": 4,
        }
      }
    >
      R$ 
      1
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "opacity": 1,
            },
          ]
        }
        testID="increase"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 0,
            }
          }
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "uri": "imageurl",
        }
      }
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
  </View>
</View>
`;
